netlink-packet/src/sock_diag/buffer.rs

```diff
@@ -0,0 +1,682 @@
+    }
+
+    pub fn set_expires(&mut self, expires: u32) {
+        let data = self.buffer.as_mut();
+        NativeEndian::write_u32(&mut data[IDIAG_RES_EXPIRES], expires)
+    }
+
+    pub fn set_rqueue(&mut self, rqueue: u32) {
+        let data = self.buffer.as_mut();
+        NativeEndian::write_u32(&mut data[IDIAG_RES_RQUEUE], rqueue)
+    }
+
+    pub fn set_wqueue(&mut self, wqueue: u32) {
+        let data = self.buffer.as_mut();
+        NativeEndian::write_u32(&mut data[IDIAG_RES_WQUEUE], wqueue)
+    }
+
+    pub fn set_uid(&mut self, uid: u32) {
+        let data = self.buffer.as_mut();
+        NativeEndian::write_u32(&mut data[IDIAG_RES_UID], uid)
+    }
+
+    pub fn set_inode(&mut self, inode: u32) {
+        let data = self.buffer.as_mut();
+        NativeEndian::write_u32(&mut data[IDIAG_RES_INODE], inode)
+    }
+}
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct RtaIterator<T> {
+    position: usize,
+    buffer: T,
+}
+
+impl<T> RtaIterator<T> {
+    pub fn new(buffer: T) -> Self {
+        RtaIterator {
+            position: 0,
+            buffer,
+        }
+    }
+}
+
+const RTA_ALIGNTO: usize = 4;
+const RTA_HDR_LEN: usize = mem::size_of::<u16>() * 2;
+
+const RTA_LENGTH: Field = 0..2;
+const RTA_TYPE: Field = 2..4;
+
+impl<T: AsRef<[u8]> + ?Sized> Iterator for RtaIterator<&T> {
+    type Item = Result<InetDiagAttr, DecodeError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // rtattr are aligned on 4 bytes boundaries, so we make sure we ignore any potential padding.
+        let offset = self.position % RTA_ALIGNTO;
+        if offset != 0 {
+            self.position += RTA_ALIGNTO - offset;
+        }
+
+        let data = self.buffer.as_ref();
+
+        if self.position >= data.len() || data.len() < RTA_HDR_LEN {
+            return None;
+        }
+
+        let data = &data[self.position..];
+        let len = NativeEndian::read_u16(&data[RTA_LENGTH]) as usize;
+        let ty = unsafe { mem::transmute(NativeEndian::read_u16(&data[RTA_TYPE])) };
+
+        if len >= data.len() {
+            return None;
+        }
+
+        let payload = &data[RTA_HDR_LEN..len];
+
+        trace!(
+            "parse {:?} extension at {} with {} bytes: {:?}",
+            ty,
+            self.position,
+            payload.len(),
+            payload
+        );
+
+        self.position += len;
+
+        Some(InetDiagAttr::parse(ty, payload))
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum InetDiagAttr {
+    MemInfo(MemInfo),
+    Info(TcpInfo),
+    Conf(String),
+    Tos(u8),
+    TClass(u8),
+    SkMemInfo(SkMemInfo),
+    Shutdown(Shutdown),
+    Protocol(u8),
+    SkV6Only(bool),
+    Mark(u32),
+    ClassId(u32),
+    Other(extension, Vec<u8>),
+}
+
+impl InetDiagAttr {
+    pub fn parse(ty: extension, payload: &[u8]) -> Result<Self, DecodeError> {
+        Ok(match ty {
+            INET_DIAG_MEMINFO if payload.len() >= mem::size_of::<MemInfo>() => {
```

Comment:
the `#[repr(C)]` will use the longest field to do the alignment, [for more detail](https://doc.rust-lang.org/std/mem/fn.size_of.html#size-of-reprc-items)

> The C representation for items has a defined layout. With this layout, the size of items is also stable as long as all fields have a stable size.

```rust
use std::mem;

#[repr(C)]
struct FieldStruct {
    first: u8,
    second: u16,
    third: u8
}

// The size of the first field is 1, so add 1 to the size. Size is 1.
// The alignment of the second field is 2, so add 1 to the size for padding. Size is 2.
// The size of the second field is 2, so add 2 to the size. Size is 4.
// The alignment of the third field is 1, so add 0 to the size for padding. Size is 4.
// The size of the third field is 1, so add 1 to the size. Size is 5.
// Finally, the alignment of the struct is 2 (because the largest alignment amongst its
// fields is 2), so add 1 to the size for padding. Size is 6.
assert_eq!(6, mem::size_of::<FieldStruct>());

```

https://doc.rust-lang.org/reference/type-layout.html

Reply:
Thanks for the explanation.